socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,773 @@
+            return 'stop'
+
+
+class ControlStateType(enum.Enum):
+    NONE = enum.auto()
+    PID_FREQ = enum.auto()
+    WAIT_FOR_TARGET_FREQ = enum.auto()
+    BRAKE = enum.auto()
+    PMX_OFF = enum.auto()
+    CONST_VOLT = enum.auto()
+    DONE = enum.auto()
+    ERROR = enum.auto()
+
+
+class ControlState:
+    """
+    Representation of the HWP Spin State.
+
+    Args
+    --------------
+    state_type : SpinStateType
+        Current state type
+    target_freq : float
+        Target Frequency for any state that changes or monitors the frequency
+    freq_thresh : float
+        Maximum difference between target freq and PID freq for the frequency to
+        be considered set correctly.
+    freq_thresh_duration : float
+        Time (sec) required within freq_thresh before moving onto DONE state.
+    voltage : float
+        Voltage to set for CONST_VOLT state
+
+    Attributes:
+    ---------------
+    success : bool, optional
+        In the DONE state, this will be True / False depending on whether
+        operation was successful. In all other states, this will be None.
+    msg : str, optional
+        Message that can be used to better describe DONE state.
+    idx : int
+        Global index of state. This will be incremented each time a SpinState
+        is instantiated, allowing you to check whether the current state is the
+        same as one you set before.
+    """
+    _idx = 0
+    _idx_lock = threading.Lock()
+
+    def __init__(self, state_type, target_freq=None, freq_thresh=None,
+                 freq_thresh_duration=None, voltage=None, pid_dir=None):
+        self.state_type = state_type
+        self.target_freq = target_freq
+        self.freq_thresh = freq_thresh
+        self.freq_thresh_duration = freq_thresh_duration
+        self.voltage = voltage
+        self.pid_dir = pid_dir
+        self.success = None
+        self.msg = None
+        self.freq_within_thresh_start = None
+        self.default_timeout = 10
+
+        self.log = txaio.make_logger()
+
+        with self._idx_lock:
+            self.idx = self._idx
+            self._idx += 1
+
+        # Prevent you from setting bad states
+        if self.state_type == ControlStateType.PID_FREQ:
+            if self.target_freq is None:
+                raise ValueError("Frequency must be set for PID_FREQ type")
+            if self.pid_dir is None:
+                raise ValueError("PID dir must be set for PID_FREQ type")
+            if self.freq_thresh is None:
+                raise ValueError("Frequency Threshold must be set for PID_FREQ type")
+            if self.freq_thresh_duration is None:
+                raise ValueError("Freq Threshold Duration must be set for "
+                                 "PID_FREQ type")
+
+        if self.state_type == ControlStateType.CONST_VOLT:
+            if self.voltage is None:
+                raise ValueError("Voltage must be set for CONST_VOLT type")
```

Comment:
You might want to add a catch for not setting `self.pid_dir` here as well. Even if the PID is not in use to control the frequency, it is still used to control the rotation direction